### DISCUSSION TITLE
Is Batect the right tool for me? (also Batect vs. Earthly)

### DISCUSSION THREAD

**Discussion by mpdude** [batect/batect#1130] [General]
👋🏼 Hi there,

this is my first issue here at Batect.

I am currently looking for ways to switch our development workflows to containers. It seems Batect and also Earthly (https://earthly.dev) are two interesting candidate tools for this.

I have a few items on my wishlist:

* It should be easy to use different tools (in different, per-project specified versions) in a project, by running all tools inside containers. 

So, no need for complex dev-VM setups, dedicated VMs per project, incompatibilities between packages (only one version can be installed at a time by the OS'es package manager) and so on. Use container-based versions of `node`, `php`, your QA tools etc. in the versions declared by the project.

* Avoid maintaining different build/orchestration files

I've seen projects that require a `Dockerfile` for building the image that goes to production, and an additional `Dockerfile.dev` that builds an image that can be used during development (maybe with extra tooling baked in, or alternative configs?). In one case, the `entrypoint.sh` in the `.dev` version of the image would take care of fetching vendors, compiling assets etc. – whereas the "production" version would do this during the build phase, specified in the `Dockerfile`.

Speaking of assets/vendors, during development you often want to mount the current project directory into the container running your server (or whatever). But when you run steps like `yarn install` in a container, you also want the results to show up in your local filesystem so your IDE can inspect the vendors/assets. 

The bottom line is: Can the tool support a style of workflow where I can get away with a _single_, comprehensive definition of how to do all this? 

#### Batect

Batect seems to be great at at least the first requirement. A great example is the tutorial where you can easily drop into a `node` container "shell". 

Also, you can easily run something like `./batect yarn -- add some-dependency`, i. e. do one-off commands with a containerized tool, while at the same time hiding complex `docker` arguments for mounting the project directory, setting the user etc.

But how do you use Batect to build the final Docker image that goes to production, or use that image and run end-to-end tests against it locally?

#### Earthly

Earthly seems to be more geared towards creating complex Docker image build processes. It can use different containers (I don't know if by _running_ or by _building_ them) to generate "artifacts", and has support for copying these artifacts between builds, containers and the local file system.

That may be beneficial by making the build process itself more portable or reproducible, but seems to fall short on the first topic of running ad hoc jobs. For example, I haven't found a way to run something like an "interactive shell" in one of the containers while at the same time having your project directory mounted into that container.

 So, maybe to conclude with a question: Does anyone have open-source examples of how they use Batect _not only_ to run tools during development, but also to produce a final Docker image of the application?

Looking forward to your thoughts.

**Comment by mpdude** [comment]
Note to self: Have a look at this.

https://github.com/batect/batect-sample-java/blob/800da38cd4b8c567be0e54a1cecdb3c7f79a4146/batect.yml#L98-L108

**Comment by charleskorn** [comment]
Thanks for the great discussion starter @mpdude. 

I'm of course biased, but I think Batect is a good fit for the problems you describe. 

I think you've already covered your first point, so I'll focus on your question on the second point:

> But how do you use Batect to build the final Docker image that goes to production, or use that image and run end-to-end tests against it locally?

You've already found the Java sample project, which is a good example of this. There are also two services that provide [update notifications](https://github.com/batect/updates.batect.dev) and [telemetry collection](https://github.com/batect/abacus) for Batect that use the same technique if you'd like other examples.

These all follow the same pattern:

- The final binary or JAR is built through a  Batect task called `build`, which places this in the build context directory of the final production image (eg. [`.batect/international-transfers-service`](https://github.com/batect/batect-sample-java/tree/main/.batect/international-transfers-service)
- End to end tests (eg. `journeyTest` in the Java sample project) then use that production image as part of the set of containers started during testing
- The image that is pushed to the image registry is built from the same directory and exploits Docker's image cache to make this effectively a no-op and get the same image that was built previously for the tests. I also know of others that want stronger guarantees about this and so use the `--tag-image` flag to tag the image built for the tests, `--override-image` to force subsequent tasks to use exactly that image without rebuilding and then push that image to the registry. 


Hope that answers your question, let me know if you'd like any clarification or if you have further questions.

**Comment by mpdude** [comment]
So, essentially, I think that is the "bind-mount volume approach" described at https://matthiasnoback.nl/2017/04/docker-build-patterns/, where you _run_ a container to build your artifacts, and use a volume mount (possible only at container runtime) to get the results out of the container.

But how does that relate to multi-platform builds? First coworkers are about to get M1 based gear in the near future, so definitely that's something I need to keep in mind.

To my understanding, multi-platform Docker builds work basically by sending the build context to different builders (running in containers? I dunno...) that work through the `Dockerfile`, and the various resulting images are combined into a multi-platform image. Volume mounts are not an option at build time (the build may even happen remote), and so the only way to get results out of this process is to copy the files out of the image.

Am I making this too much of a problem? Do you see any challenges arising from that for a purely Batect-based approach?

**Reply by charleskorn** [reply]
> So, essentially, I think that is the "bind-mount volume approach" described at https://matthiasnoback.nl/2017/04/docker-build-patterns/, where you run a container to build your artifacts, and use a volume mount (possible only at container runtime) to get the results out of the container.

Yep, that's correct.

> But how does that relate to multi-platform builds? First coworkers are about to get M1 based gear in the near future, so definitely that's something I need to keep in mind.
> 
> To my understanding, multi-platform Docker builds work basically by sending the build context to different builders (running in containers? I dunno...) that work through the Dockerfile, and the various resulting images are combined into a multi-platform image. Volume mounts are not an option at build time (the build may even happen remote), and so the only way to get results out of this process is to copy the files out of the image.
> 
> Am I making this too much of a problem? Do you see any challenges arising from that for a purely Batect-based approach?

Multi-platform builds make life a little more complicated, yes :) It also depends on your tech stack a little as well. For example, if you're building an app for the JVM, then you should be able to just build the JAR once and use that to build platform-specific images as you need. In contrast, for something like Golang, if you were to follow the same pattern, your build task would need to build binaries for x86 and ARM, and then your production image build would need to pick the right binary for the architecture being built. 

In either group, something like a multi-platform image build would work. You might also be able to get away with only building for the local platform (eg. x86 on an Intel Mac, ARM on a M1) which might simplify things at the risk of only surfacing issues on either platform when the build is run on that platform - for example, someone on a M1 might break things for x86 and not know about it until someone on x86 tries to build their changes.
